server/src/main/java/org/opensearch/tiering/TieringServiceValidator.java

```diff
@@ -0,0 +1,334 @@
+    }
+
+    /**
+     * Validates the health of the specified index in the current cluster state.
+     *
+     * @param currentState the current cluster state
+     * @param index the index to be validated
+     * @return true if the index health is not in the "red" state, false otherwise
+     */
+    public static boolean validateIndexHealth(final ClusterState currentState, final Index index) {
+        final IndexRoutingTable indexRoutingTable = currentState.routingTable().index(index);
+        final IndexMetadata indexMetadata = currentState.metadata().index(index);
+        final ClusterIndexHealth indexHealth = new ClusterIndexHealth(indexMetadata, indexRoutingTable);
+        return !ClusterHealthStatus.RED.equals(indexHealth.getStatus());
+    }
+
+    /**
+     * Validates that the specified index is in the open state in the current cluster state.
+     *
+     * @param currentState the current cluster state
+     * @param index the index to be validated
+     * @return true if the index is in the open state, false otherwise
+     */
+    public static boolean validateOpenIndex(final ClusterState currentState, final Index index) {
+        return currentState.metadata().index(index).getState() == IndexMetadata.State.OPEN;
+    }
+
+    /**
+     * Validates that the disk threshold low watermark is not breached on all the eligible nodes in the cluster.
+     *
+     * @param currentState the current cluster state
+     * @param clusterInfo the current nodes usage info for the cluster
+     * @param diskThresholdSettings the disk threshold settings of the cluster
+     * @throws IllegalArgumentException if the disk threshold low watermark is breached on all eligible nodes
+     */
+    public static void validateDiskThresholdWaterMarkNotBreached(
+        final ClusterState currentState,
+        final ClusterInfo clusterInfo,
+        final DiskThresholdSettings diskThresholdSettings
+    ) {
+        final Map<String, DiskUsage> usages = clusterInfo.getNodeLeastAvailableDiskUsages();
+        if (usages == null) {
+            logger.trace("skipping monitor as no disk usage information is available");
+            return;
+        }
+        final Set<String> nodeIds = getEligibleNodes(currentState).stream().map(DiscoveryNode::getId).collect(Collectors.toSet());
+        int count = 0;
+        for (String node : usages.keySet()) {
+            if (nodeIds.contains(node)) {
+                if (usages.get(node).getFreeBytes() < diskThresholdSettings.getFreeBytesThresholdLow().getBytes()) {
+                    logger.warn("Disk threshold low watermark breached on the node [{}]", node);
+                    count++;
+                }
+            }
+        }
+
+        if (count == nodeIds.size()) {
+            throw new IllegalArgumentException(
+                "Disk threshold low watermark breached on all the search nodes, hence no new index can be accepted for tiering "
+            );
+        }
```

Comment:
this would throw an exception even when Disk threshold low watermark is breached on one of the warm nodes, thats not what we want

Reply:
there is a `return` statement in for loop inside `if` check. As soon as a node is found to not breach watermark it will return from the method. So exception will be thrown only in case when all nodes breach the watermark.

Reply:
got it, let me update it